upgrade/src/confirmTransaction.js

```diff
@@ -0,0 +1,39 @@
+const confirmTransaction = async ({ contract, fromBlock = 0, destination, data, address, gasPrice }) => {
+  const submissions = await contract.getPastEvents('Submission', { fromBlock, toBlock: 'latest' })
+
+  if (!submissions.length) {
+    throw new Error('No submissions found.')
+  }
+  const filteredSubmission = []
+  await Promise.all(
+    submissions.map(async s => {
+      const { transactionId } = s.returnValues
+      const transaction = await contract.methods.transactions(transactionId).call()
+      if (transaction.destination === destination && transaction.data === data) {
+        if (!transaction.executed) {
+          console.log(
+            `Found a transaction to confirm. Id: ${transactionId}, Block: ${s.blockNumber}, Data: ${transaction.data}`
+          )
+          filteredSubmission.push(s)
+          return
+        }
+        console.log(
+          `Found a transaction that was already executed. Id: ${transactionId}, Block: ${s.blockNumber}, Data: ${transaction.data}`
+        )
+      }
+    })
+  )
+
+  if (filteredSubmission.length > 1) {
+    throw new Error('More than one transaction found.')
+  } else if (filteredSubmission.length === 0) {
+    throw new Error('No transaction to confirm.')
+  }
+  const { transactionId } = filteredSubmission[0].returnValues
+
+  const gas = await contract.methods.confirmTransaction(transactionId).estimateGas({ from: address })
+  const receipt = await contract.methods.confirmTransaction(transactionId).send({ from: address, gas, gasPrice })
```

Comment:
Imagine that all validators performs a particular confirmation at the same time. The last confirmation will consume more gas than it was estimated since the transaction will not only change the state of the multisig wallet contract but also will invoke the method in a destination contract. That's why in the instructions (e.g. https://docs.tokenbridge.net/xdai-bridge/xdai-bridge-contracts-management/upgrade-contracts, the step 6) suggest to multiply the gas usage parameter for the confirmation transaction. Is it worth to reflect this behaviour here?

Reply:
Yes, it makes sense to do it, updated the logic to multiply the estimated gas by 4 times. https://github.com/poanetwork/tokenbridge-contracts/pull/337/commits/a3410d1ec1e4280b98412b19c8fc44020fc5d17b